util.cpp

```diff
@@ -0,0 +1,55 @@
+#include <stdexcept>
+#include <boost/multiprecision/cpp_int.hpp>
+#include "include/util.h"
+
+using uint128_t = boost::multiprecision::uint128_t;
+
+const ull ULLMAX = std::numeric_limits<ull>::max();
+
+unsigned long long int double_to_ull(double d, double epsilon) {
+  return (unsigned long long) (d + epsilon);
+}
+
+ull nondirectional_non_self_edge_pairing_fn(ull i, ull j) {
+  // swap i,j if necessary
+  if (i > j) {
+    i^=j;
+    j^=i;
+    i^=j;
+  }
+  ull jm = j-1ull;
+  if ((j & 1ull) == 0ull) j>>=1ull;
+  else jm>>=1ull;
+  if (ULLMAX/j < jm)
+    throw std::overflow_error("Computation would overflow unsigned long long max");
+  j*=jm;
+  if (ULLMAX-j < i)
+    throw std::overflow_error("Computation would overflow unsigned long long max");
+  return i+j;
+}
+
+std::pair<ull, ull> inv_nondir_non_self_edge_pairing_fn(ull idx) {
```

Comment:
It's unclear to me what this function is doing exactly - think about this for a bit and see if you can write it in a clearer way

Reply:
this inverts the pairing function. I'm not too sure I could make the calculations more transparent. I could add the explicit calculations into the header/documentation if you think it would be helpful

Reply:
No this is okay - if you have some reference for what this pairing is doing or where it's taken from you could include it in a comment (like a citation almost) which would help future coders figure out what you were doing